exercises/enumerating_trees/descr.md

```diff
@@ -0,0 +1,364 @@
+the empty set.
+
+**Question 2.** Define a function `just` of type `'a -> 'a enum` such that
+`just x` represents a singleton set, whose element is `x`, and the size of
+this element is considered to be zero.
+
+**Question 3.** Define a function `pay` of type `'a enum -> 'a enum` such that
+the enumeration `pay e` represents the same set as the enumeration `e`, but
+the size of every element is increased by one. That is, if a value `x` appears
+in `e` among the elements of size `s`, then it should appear in `pay e` among
+the elements of size `s+1`.
+
+**Question 4.** Define a function `sum` of type `'a enum -> 'a enum -> 'a
+enum` such that `sum e1 e2` represents the union of the sets `e1` and `e2`.
+One can assume that the sets `e1` and `e2` are disjoint.
+
+**Question 5.** Define a function `product` of type
+`'a enum -> 'b enum -> ('a * 'b) enum`
+such that `product e1 e2` represents the Cartesian product
+of the sets `e1` and `e2`.
+By convention, the size of a pair `(x1, x2)` is the sum of the sizes
+of `x1` and `x2`.
+
+*Hint.* A pair `(x1, x2)` has size `s` if and only if
+`x1` has size `s1` and `x2` has size `s2`
+*for some* integers `s1` and `s2` such that `s1 + s2 = s` holds.
+
+*Hint.* Define an auxiliary function `up: int -> int -> int list`
+such that `up i j` is the list `[i; i+1; ...; j]`.
+Use the functions `Seq.bigsum`, `List.map`, and `Seq.product` so
+as to combine all of the possibilities that arise out of
+all valid choices of `s1` and `s2`.
+
+**Question 6.** Define a function `map` of type
+`('a -> 'b) -> 'a enum -> 'b enum`
+such that the set `map f e` is the image through `f` of the set `e`.
+By convention, the size of the element `f x` in the enumeration `map f e`
+is the size of the element `x` in the enumeration `e`.
+
+## Enumerating lists
+
+In the following, we use the constructor functions defined above.
+
+**Question 7.** Define an enumeration `bit` of type `int enum`
+whose values, `0` and `1`, are both considered to have size 0.
+
+We assume that a memoizing fixed point combinator `fix` is given. Its type is
+`(('a -> 'b) -> ('a -> 'b)) -> ('a -> 'b)`. The idea is, instead of defining
+an ordinary recursive function by `let rec f x = e`, a user can define a
+memoizing recursive function by `let f = fix (fun f x -> e)`.
+The implementation of `fix` uses a hash table;
+there is an implicit assumption that
+OCaml's generic equality and hash functions
+can be used at type `'a`.
+For more details, see the exercise "Counting Trees".
+
+It is worth noting that, because an enumeration is a function,
+`fix` also has type
+`('a enum -> 'a enum) -> 'a enum`
+as a special case.
+Thus, `fix` can be used to recursively define an enumeration.
+
+**Question 8.** Define a function `list` of type `'a enum -> 'a list enum` such
+that, if `e` is an enumeration of a certain set of elements, then `list e` is
+an enumeration of the lists of elements drawn from `e`. By convention, the
```

Comment:
`list` must force the evaluation of `e 0; e 1; ...etc`, right? Can you motivate the need for this function? It seems inconsistent with the previous functions that are lazy.

Reply:
I don't see what you mean. Calling `list e` does not do much computation; it only allocates a closure, say, `l`, which it returns. Then, when `l` is applied to some size `s`, then probably this will cause `e i` to be invoked for every `i` less than or equal to `s`. It is probably a good thing if the function `e` is memoized, as `e i` can be called more than once for a given `i`.

Reply:
Ignore this comment: I read `a enum list` instead of `a list enum`.